### PR TITLE
LUKS2: add AEAD integrity option

### PR DESCRIPTION
**Pull request by aplanas**
LUKS2 support "integrity" and "data authentication" since 2.0 in
experimental mode.  This is required when we want to detect changes out
of band in the devices.

Recently the kernel is supporting good AEAD family of algorithms that
will guarantee unauthorized changes in the device (instead of decrypting
random data) for big disks.

For integrity it is recommended "aegis128-random" with AEAD, as other
family (like chacha20-random with poly1305, or aes-xts-plain64 with
hmac-sha256) have performance problems, or are not supporting large
enought devices.  A check is implemented to validate the data.

This patch add a new element, "integrity", that for now can only have
"aead" value.  When other methods are added, the check can be changed
without changing the XML data, nor the parser.  If integrity is set,
also it is checked that "aegis128-random" cipher algorithm is used.

Signed-off-by: Alberto Planas <aplanas@suse.com>

### CODE REVIEW

**Comment by aschnell** [comment]
Is the resulting device with integrity smaller than without? In that case the space calculation must be extended.

File: storage/Devices/Encryption.h [hunk 1]
```diff
@@ -143,6 +143,24 @@ namespace storage
 	 */
 	void set_pbkdf(const std::string& pbkdf);
 
+	/**
+	 * Get the integrity. For now we only expect AEAD.
+	 *
+	 * Currently only supported for LUKS2.
```
Comment (aschnell) [inline storage/Devices/Encryption.h:149]: Please add that it is even experimental in cryptsetup/luks.

Comment (aplanas) [reply storage/Devices/Encryption.h:149]: done

**Comment by aplanas** [comment]
> Is the resulting device with integrity smaller than without? In that case the space calculation must be extended.

Do I need to implement some method? Can you point me to some example in the code?

Would be OK if the estimation is an approximation?

**Comment by aplanas** [comment]
> Luks::Impl::calculate_region_and_topology()

Thanks. Should be done.

**Comment by aplanas** [comment]
> Please add a testcase for probing in testsuite/probe.

I will try to do. So far there are no version 2 luks tests there, so I need to create all the fixtures from scratch.

File: storage/Devices/LuksImpl.cc [hunk 2]
```diff
@@ -603,6 +617,14 @@ namespace storage
 		ST_THROW(Exception("invalid encryption type"));
 	}
 
+	if (get_cipher() == "aegis128-random" || !get_integrity().empty()) {
+	    if (get_cipher() != "aegis128-random")
+		ST_THROW(Exception("when integrity is set it is expected aegis128-random cipher"));
+
+	    if (get_integrity() != "aead")
+		ST_THROW(Exception("for now only AEAD integrity is supported"));
+	}
```
Comment (aschnell) [inline storage/Devices/LuksImpl.cc:626]: Checking this during commit is very late. Likely a better place is in check().

And please add a document explaining the situation (what works, what does not work, what kernel is needed, ...) in the doc directory.

File: storage/SystemInfo/CmdCryptsetup.h [hunk 3]
```diff
@@ -103,6 +104,11 @@ namespace storage
 	 */
 	string pbkdf;
 
+	/**
+	 * Is integrity used. Currently is only expected AEAD. Only fir LUKS2.
```
Comment (aschnell) [inline storage/SystemInfo/CmdCryptsetup.h:108]: Typo.

Comment (aplanas) [reply storage/SystemInfo/CmdCryptsetup.h]: Fixed.

**Comment by aplanas** [comment]
> If probing works you can just save the mockup and devicegraph using utils/probe --save-mockup --save-devicegraph.

Done. There is a new test in prob using this process (integrity-*)

I think that I addressed all the comments.

**Comment by aschnell** [comment]
cryptsetup works for me only when also setting --key-size 128. If that is expected please document it. And create a integration test which does test the creation.

**Comment by aplanas** [comment]
> cryptsetup works for me only when also setting --key-size 128. If that is expected please document it.

Right. I will add it.

> And create a integration test which does test the creation.

I will check how to do that.

**Comment by aschnell** [comment]
There is also a problem with probing due to the extra dm table. If the LUKS device is open the probed name includes the "_dif". Here the code Luks::Impl::probe_lukses() must check the target and if it is "integrity" find the table with target "crypt".

The probe test right now does not check this since the LUKS device is not open.

**Comment by aschnell** [comment]
Is QA going to test this? And reference to a feature request?

**Comment by aplanas** [comment]
> Is QA going to test this?

I cannot answer this, but I guess so.

> And reference to a feature request?

It is part of FDE and PM-3419

**Comment by aplanas** [comment]
> There is also a problem with probing due to the extra dm table. If the LUKS device is open the probed name includes the "_dif". Here the code Luks::Impl::probe_lukses() must check the target and if it is "integrity" find the table with target "crypt".

I will check what it this is about. But did not the added test with the mocks was checking this?

> The probe test right now does not check this since the LUKS device is not open.

File: integration-tests/luks/create-with-integrity.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+environment = Environment(False)
+
+storage = Storage(environment)
+storage.probe()
+
+staging = storage.get_staging()
+
+sdc1 = Partition.find_by_name(staging, "/dev/sdc1")
+
+luks = to_luks(sdc1.create_encryption("cr-test"))
```
Comment (aschnell) [inline integration-tests/luks/create-with-integrity.py:21]: The create_encryption function without providing the encryption type is deprecated.

Comment (aplanas) [reply integration-tests/luks/create-with-integrity.py]: Done.

File: storage/Devices/LuksImpl.cc [hunk 5]
```diff
@@ -400,7 +419,31 @@ namespace storage
...
+			if (table.target == "integrity")
+			    integrity = true;
+		    }
+
+		    if (integrity)
+		    {
+			dm_table_name = dm_table_name.substr(0, dm_table_name.length() - 4);
+			const CmdDmsetupInfo& cmd_dmsetup_info = system_info.getCmdDmsetupInfo();
+			if (!cmd_dmsetup_info.exists(dm_table_name))
+			    ST_THROW(Exception("integrity device found, missing the associated crytpo target"));
```
[the remark this answers is not in the document]
Comment (aplanas) [reply storage/Devices/LuksImpl.cc:443]: Not sure to understand. "crypt" is the name in the mockup, but can be anything. If there is an integrity device with _dif, there must be an associated device with target "crypt", that is why was like that before.

File: storage/Devices/LuksImpl.cc [hunk 6]
```diff
@@ -400,7 +419,31 @@ namespace storage
...
+			if (table.target == "integrity")
+			    integrity = true;
+		    }
+
+		    if (integrity)
+		    {
+			dm_table_name = dm_table_name.substr(0, dm_table_name.length() - 4);
+			const CmdDmsetupInfo& cmd_dmsetup_info = system_info.getCmdDmsetupInfo();
+			if (!cmd_dmsetup_info.exists(dm_table_name))
+			    ST_THROW(Exception("integrity device found, missing the associated 'crytp' target"));
```
Comment (aschnell) [inline storage/Devices/LuksImpl.cc:443]: I am just talking about the typo here: "crytp" -> "crypt"

Comment (aplanas) [reply storage/Devices/LuksImpl.cc]: I am blind. Done. (Sorry)

**Comment by aplanas** [comment]
> Please update the probe test to have the LUKS device open so that the new code in Luks::Impl::probe_lukses is used.

Right. Done.

File: storage/Devices/LuksImpl.cc [hunk 7]
```diff
@@ -96,6 +107,14 @@ namespace storage
 	calculate_region_and_topology();
     }
 
+    void
+    Luks::Impl::set_integrity(const string& integrity)
+    {
+	Encryption::Impl::set_integrity(integrity);
+
+	if (get_integrity() != integrity)
```
Comment (aschnell) [inline storage/Devices/LuksImpl.cc:115]: Is that ever the case?

Comment (aplanas) [reply storage/Devices/LuksImpl.cc:115]: In reality no, but maybe someone put manually this suffix?

File: storage/Devices/LuksImpl.cc [hunk 8]
```diff
@@ -496,6 +541,10 @@ namespace storage
 	else
 	    size = 0 * B;
 
+	// If we use integrity, calculate a pesimistic size
+	if (!get_integrity().empty())
+	    size = static_cast<unsigned long long>(size * 0.85);
```
Comment (aschnell) [inline storage/Devices/LuksImpl.cc:546]: Why not ```size *= 0.85```?

Comment (aplanas) [reply storage/Devices/LuksImpl.cc:546]: To be more explicit in the type, but I will change it (not used to C++ idioms)

File: storage/Devices/LuksImpl.cc (hunk 7 above)
Comment (aschnell) [reply storage/Devices/LuksImpl.cc:115]: After ```set_integrity(integrity)``` the condition is always false AFAIS.

Comment (aplanas) [reply storage/Devices/LuksImpl.cc:115]: Ah right, this will never call the reize. I will move the condition before.

Comment (aplanas) [reply storage/Devices/LuksImpl.cc:113]: Done. If you think that the condition should be dropped, we need to figure out why the tests are falling.

File: storage/Devices/LuksImpl.cc (hunk 8 above)
Comment (aplanas) [reply storage/Devices/LuksImpl.cc]: Done

File: storage/Devices/Encryption.h [hunk 9]
```diff
@@ -143,6 +143,24 @@ namespace storage
...
+	 * Get the integrity. For now we only expect AEAD.
+	 *
+	 * Currently only supported for LUKS2 (experimental).
+	 */
+	const std::string& get_integrity() const;
+
+	/**
+	 * Set the integrity. If the integrity is set, this will
+	 * restrict the number of allowed cipher algorithms.  Also the
+	 * expected value in "aead".  Technically this can be replaced
```
Comment (aschnell) [inline storage/Devices/Encryption.h:156]: "in" -> "is". Empty is also allowed.

File: storage/Devices/Encryption.h [hunk 10]
```diff
@@ -143,6 +143,24 @@ namespace storage
...
+	 * as a bootlean flag, but we maintain a string in case of
```
Comment (aschnell) [inline storage/Devices/Encryption.h:157]: "bootlean" -> "boolean".

File: storage/Devices/Encryption.h [hunk 11]
```diff
@@ -143,6 +143,24 @@ namespace storage
...
 
+	/**
+	 * Get the integrity. For now we only expect AEAD.
+	 *
+	 * Currently only supported for LUKS2 (experimental).
+	 */
+	const std::string& get_integrity() const;
+
+	/**
+	 * Set the integrity. If the integrity is set, this will
```
Comment (aschnell) [inline storage/Devices/Encryption.h:154]: Like other functions (set_pbkdf) it should mention that the value is "Only used during creation.".

Comment (aplanas) [reply storage/Devices/Encryption.h:154]: Done

File: storage/Devices/Encryption.h (hunk 10 above)
Comment (aplanas) [reply storage/Devices/Encryption.h]: Fixed.

File: storage/Devices/Encryption.h (hunk 9 above)
Comment (aplanas) [reply storage/Devices/Encryption.h]: Fixed.

File: storage/Devices/LuksImpl.cc (hunk 7 above)
Comment (aschnell) [reply storage/Devices/LuksImpl.cc:113]: Likely a check like in ```Md::Impl::calculate_region_and_topology()``` is needed:

```
	if (exists_in_system())
	    return;
```

Explanation is in Md::Impl.

Comment (aplanas) [reply storage/Devices/LuksImpl.cc]: Added check in local `calculate_region_and_topology()`

**Comment by aschnell** [comment]
Looks good now. Please add the feature request in the git message and increase version in LIBVERSION to .1.74.0.

**Comment by aplanas** [comment]
> Looks good now. Please add the feature request in the git message and increase version in LIBVERSION to 1.74.0.

Should be done.

**Pull request merged**